### PR TITLE
OH and Binary Encoders weighting fix

### PR DESCRIPTION
**Pull request by George3d6**
What's the bug here?

The One hot encoder and the binary were **inverting** the target weights argument when creating the weights-per-index tensor (used by the Neural mixer).

Why was this not affecting accuracy?

- Because we never weighted manually, only automatically, and instead of passing the *reverse* occurrence of every class to the encoders we were passing the occurrence of every class ... so double investing => all is good

This fix means that the `target_weight` argument now works with the `Neural` encoder *and* might increase accuracy in the case where implicit weights were passed to LightGBM, since the latest release it takes weight into account.

I'm actually surprised we didn't see more of an accuracy drop due to passing the wrong weights to LGBM in the previous PR, but I guess it shows that Neural is able to pick up the slack... and probably that we need a better benchmarks methodology.

### CODE REVIEW

File: lightwood/encoder/categorical/onehot.py [hunk 1]
```diff
@@ -79,7 +79,8 @@ def prepare(self, priming_data, max_dimensions=20000):
                 self.target_weights[UNCOMMON_WORD] = uncommon_weight
             for word in set(priming_data):
                 if self.target_weights is not None:
-                    self.index_weights[self._lang.word2index[str(word)]] = 1 / self.target_weights[word]
+                    self.index_weights[self._lang.word2index[str(word)]] = \
+                        self.target_weights[word] / np.max(list(self.target_weights.values()))
```
Comment (hakunanatasha) [inline lightwood/encoder/categorical/onehot.py:83]: see above; you will accidentially down-weight minority classes

**Review by hakunanatasha — CHANGES_REQUESTED** [review]
can you double check your logic around weighting? Index_weights is supposed to represent the inverse weights (Which is fine). You have right now something that will continue to exacerbate minority issues because it will do class_counts_of_class_i / max(class_counts)

File: lightwood/api/json_ai.py [hunk 2]
```diff
@@ -119,7 +119,7 @@ def lookup_encoder(
             if problem_defintion.unbias_target:
                 encoder_dict["args"][
                     "target_weights"
-                ] = "$statistical_analysis.target_class_distribution"
+                ] = "$statistical_analysis.assumed_target_class_weight"
```
[the remark this answers is not in the document]
Comment (George3d6) [reply lightwood/api/json_ai.py:122]: because it's the inverse frequency,  but even then, we might want to have another technique for weighting (e.g. something on a log scale later)

File: lightwood/api/types.py [hunk 3]
```diff
@@ -174,6 +174,7 @@ class StatisticalAnalysis:
     df_target_stddev: Optional[float]
     train_observed_classes: object  # Union[None, List[str]]
     target_class_distribution: object  # Dict[str, float]
```
[the remark this answers is not in the document]
Comment (George3d6) [reply lightwood/api/types.py:176]: See comment above, this is the default weighting for classes, which in the future we might actually want to be something else, at least that was my thinking.

File: lightwood/api/types.py [hunk 4]
```diff
@@ -174,6 +174,7 @@ class StatisticalAnalysis:
     df_target_stddev: Optional[float]
```
Comment (paxcema) [inline lightwood/api/types.py:174]: I think for this it would be super helpful to add the meaning/purpose of `target_class_distribution` vs `assumed_target_class_weight` to avoid user confusion

File: lightwood/api/json_ai.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (paxcema) [reply lightwood/api/json_ai.py:122]: If this is customizable by the user, then I think "assumed" is not needed in the name here.

**Review by paxcema — APPROVED** [review]
I'd be okay with this fix as long as the actual meaning of the params makes its way into the docstring of the stat analyzer.

**Comment by George3d6** [comment]
@paxcema I renamed that since both you and @hakunanatasha commented on it *and* added a docstring for clarity. Let me know if it makes more sense now.

**Comment by paxcema** [comment]
Does make sense, thanks :+1:

File: lightwood/api/json_ai.py [hunk 5]
```diff
@@ -119,7 +119,7 @@ def lookup_encoder(
             if problem_defintion.unbias_target:
                 encoder_dict["args"][
                     "target_weights"
-                ] = "$statistical_analysis.target_class_distribution"
+                ] = "$statistical_analysis.suggested_target_class_weight"
```
Comment (hakunanatasha) [inline lightwood/api/json_ai.py:122]: just call it target_class_weights versus target_class_distributions. Target_class_weights will always refer to what is passed to the model henceforth.

File: lightwood/api/types.py [hunk 6]
```diff
@@ -161,19 +161,21 @@ class StatisticalAnalysis:
     :param df_target_stddev: The standard deviation of the target of the dataset
     :param train_observed_classes:
     :param target_class_distribution:
+    :param suggested_target_class_weight: What weight the analysis suggests to assign each class by in the case of classification problems. Note: target_weights in the problem definition overides this
```
Comment (hakunanatasha) [inline lightwood/api/types.py:164]: please just call it `Target_class_weights`

File: lightwood/api/types.py (hunk 4 above)
Comment (hakunanatasha) [reply lightwood/api/types.py:175]: call distribution == number of instances of each class; and weights == assigned importance to each class (whether this is related to the weights or not doesnt matter)

File: lightwood/api/types.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hakunanatasha) [reply lightwood/api/types.py:177]: in the future we will either change this, or a user will just implement their own approach

File: lightwood/api/types.py [hunk 7]
```diff
@@ -161,19 +161,21 @@ class StatisticalAnalysis:
...
     :param avg_words_per_sentence:
     :param positive_domain:
-    """
+    """ # noqa
 
     nr_rows: int
     df_target_stddev: Optional[float]
     train_observed_classes: object  # Union[None, List[str]]
     target_class_distribution: object  # Dict[str, float]
+    suggested_target_class_weight: object  # Dict[str, float]
```
Comment (hakunanatasha) [inline lightwood/api/types.py:178]: replace to `Target_class_weights`

File: lightwood/data/statistical_analysis.py [hunk 8]
```diff
@@ -133,8 +133,12 @@ def statistical_analysis(data: pd.DataFrame,
 
     # get observed classes, used in analysis
     target_class_distribution = None
+    suggested_target_class_weight = None
```
Comment (hakunanatasha) [inline lightwood/data/statistical_analysis.py:136]: replace to `target_class_weights`

**Review by hakunanatasha — CHANGES_REQUESTED** [review]
please change 'suggested_target_weights' to just `target_class_weights` and then I will approve.

**Comment by hakunanatasha** [comment]
One note - please make an edge case/unit-test/assert that handles if a count is 0 for some reason, as this will assuredly crop up sometime.
